bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthFactoryImpl.java

```diff
@@ -0,0 +1,154 @@
+
+    private BundleContext bundleContext;
+    private OAuthStoreHandler oAuthStoreHandler;
+    int tokenExpiresIn_Buffer = OAuthClientServiceImpl.DEFAULT_TOKEN_EXPIRES_IN_BUFFER_SECOND;
+
+    private final Map<String, OAuthClientServiceImpl> oauthClientServiceCache = new HashMap<>();
+
+    @Activate
+    public void activate(BundleContext bundleContext) {
+        this.bundleContext = bundleContext;
+    }
+
+    @Deactivate
+    public void deactivate() {
+        // close each service
+        for (OAuthClientServiceImpl clientServiceImpl : oauthClientServiceCache.values()) {
+            clientServiceImpl.close();
+        }
+        oauthClientServiceCache.clear();
+    }
+
+    @Override
+    public String createOAuthClientService(@NonNull String tokenUrl, @Nullable String authorizationUrl,
+            @NonNull String clientId, @Nullable String clientSecret, @Nullable String scope,
+            @Nullable Boolean supportsBasicAuth) {
+
+        String uuidHandle = UUID.randomUUID().toString();
+
+        @SuppressWarnings("null")
+        OAuthClientServiceImpl clientImpl = OAuthClientServiceImpl.getInstance(bundleContext, uuidHandle,
+                oAuthStoreHandler, tokenUrl, authorizationUrl, clientId, clientSecret, scope, supportsBasicAuth,
+                tokenExpiresIn_Buffer);
+
+        oauthClientServiceCache.put(uuidHandle, clientImpl);
+        return uuidHandle;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public OAuthClientService getOAuthClientService(@NonNull String handle) {
+        Preconditions.checkNotNull(handle, "handle must not be null");
+
+        OAuthClientServiceImpl clientImpl = null;
+        try {
+            clientImpl = oauthClientServiceCache.get(handle);
+            if (clientImpl == null || clientImpl.isClosed()) {
+                // This happens after reboot, or client was closed without factory knowing; create a new client
+                // the store has the handle/config data
+                clientImpl = OAuthClientServiceImpl.getInstance(bundleContext, handle, oAuthStoreHandler,
+                        tokenExpiresIn_Buffer);
+                oauthClientServiceCache.put(handle, clientImpl);
+            }
+        } catch (OAuthException e) {
+            logger.error("Cannot get OAuth client service by handle {}", handle, e);
+        }
+        return clientImpl;
+    }
+
+    @Override
+    public void ungetOAuthService(@NonNull String handle) {
+        Preconditions.checkNotNull(handle, "handle must not be null");
+
+        OAuthClientServiceImpl clientImpl = oauthClientServiceCache.get(handle);
+        if (clientImpl == null) {
+            logger.error("{} handle not found.  Cannot unregisterOAuthServie", handle);
+            return;
+        }
+        clientImpl.close();
+    }
+
+    @Override
+    public void deleteServiceAndAccessToken(@NonNull String handle) {
+        Preconditions.checkNotNull(handle, "handle must not be null");
```

Comment:
since the annotation is already added, you can remove the precondition.

Reply:
The annotations work in different ways than the Preconditions. The annotation warnings can also be suppressed. I believe it is easier for developers to see which param is nullable/ nonnull with annotation. To enforce them will have to rely on the preconditions

Reply:
As far as I know, it is enforced in ESH using annotations. You can see in other bundles as well.